### PR TITLE
Support REST external reference

### PR DESCRIPTION
**Pull request by hellosnow**
@chenkennt @ansyral @vwxyzh  @vicancy @qinezh @superyyrrzz

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs [hunk 1]
```diff
@@ -50,31 +62,55 @@ private SwaggerObjectBase Build(JToken token)
...
+                            jObject.Remove("$ref");
+
+                            // Currently we support external reference without $ref inside
+                            var externalSwaggerPath = Path.Combine(swaggerDir, swaggerReference.Path);
+                            using (JsonReader reader = new JsonTextReader(File.OpenText(externalSwaggerPath)))
+                            {
+                                var externalJObject = JObject.Load(reader);
+                                foreach (var item in externalJObject)
+                                {
+                                    jObject[item.Key] = item.Value;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: >jObject[item.Key] [](start = 36, length = 17)

Can existing property be overwritten? Do we need log warning?

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs [hunk 2]
```diff
@@ -19,16 +22,25 @@ internal class SwaggerJsonBuilder
...
-            var swagger = Build(token);
-            RemoveReferenceDefinitions((SwaggerObject)swagger);
-            return ResolveReferences(swagger, new Stack<string>());
+            using (JsonReader reader = new JsonTextReader(File.OpenText(swaggerPath)))
+            {
+                _documentObjectCache = new Dictionary<string, SwaggerObjectBase>();
+                var token = JToken.ReadFrom(reader);
+
+                var swaggerDir = Path.GetDirectoryName(swaggerPath);
+                if (string.IsNullOrEmpty(swaggerDir))
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: >string.IsNullOrEmpty(swaggerDir) [](start = 20, length = 32)

relative swaggerPath is not allowed?

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs [hunk 3]
```diff
@@ -50,31 +62,55 @@ private SwaggerObjectBase Build(JToken token)
...
+                            // These properties cannot be referenced
+                            jObject.Remove("$ref");
+                            deferredObject.Token = jObject;
+                            _documentObjectCache.Add(location, deferredObject);
+                            return deferredObject;
+                        case RestApiHelper.SwaggerReferenceType.ExternalReference:
+                            jObject.Remove("$ref");
+
+                            // Currently we support external reference without $ref inside
+                            var externalSwaggerPath = Path.Combine(swaggerDir, swaggerReference.Path);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: >externalSwaggerPath [](start = 32, length = 19)

should check existence of this path

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs [hunk 4]
```diff
@@ -27,25 +32,66 @@ public static string FormatDefinitionSinglePath(string reference)
...
+                    Path = reference,
+                    Name = fileName
+                };
             }
 
-            throw new InvalidOperationException($"Full reference path \"{reference}\" must start with '/' or '#/'");
+            throw new InvalidOperationException($"Reference path \"{reference}\" is not supported now");
+        }
+
+        public class SwaggerReference
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: >SwaggerReference [](start = 21, length = 16)

move out?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: yes plan to

---
In reply to: [93162767](https://github.com/dotnet/docfx/pull/1019#discussion_r93162767) [](ancestors = 93162767)

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs (hunk 2 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: in this method the path is full path, to open read in line 27

---
In reply to: [93162486](https://github.com/dotnet/docfx/pull/1019#discussion_r93162486) [](ancestors = 93162486)

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs (hunk 1 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: In my understanding exsiting property could be overwritten, so should warning here

---
In reply to: [93157759](https://github.com/dotnet/docfx/pull/1019#discussion_r93157759) [](ancestors = 93157759)

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs (hunk 3 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: >externalSwaggerPath [](start = 32, length = 19)

if external swagger contains external reference, swaggerDir should be external swagger file's dir?

**Comment by vwxyzh** [comment]
:shipit:

**Comment by vicancy** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
:shipit:

**Pull request merged**